### ISSUE TITLE
Counting words for Asian languages

### ISSUE THREAD

**Issue by atimmer** [Yoast/YoastSEO.js#761]
WordPress has build code to be able to count words in Asian languages. We should take this as inspiration. WordPress actually ships the word counting code only in the Chinese language pack, the third link shows where to find it.

Patch for word-count.js
https://core.trac.wordpress.org/ticket/20738
https://core.trac.wordpress.org/ticket/30966

Where to find the language pack:
https://core.trac.wordpress.org/ticket/33454

**Comment by monbauza** [comment]
Please inform the customer of conversation # 99154 when this conversation has been closed.

**Comment by monbauza** [comment]
Please inform the customer of conversation # 148759 when this conversation has been closed.

**Comment by paulovsky** [comment]
Any developments on this issue?

**Comment by monbauza** [comment]
Please inform the customer of conversation # 172656 when this conversation has been closed.

**Comment by a4jp-com** [comment]
Maybe a character count is better for Japanese/Chinese. One kanji character is usually takes up the space of 2 regular characters.

**Comment by terw-dan** [comment]
@a4jp-com thanks for your suggestion. So each kanji character can be seen as 1 word? Or is there more to it?

**Comment by paulovsky** [comment]
Japanese (also Chinese, Korean, Vietnamese) is a logographic language from the Han family; that means that not all characters represent morphemes: some morphemes are composed of more than one characters (see more [here](https://en.wikipedia.org/wiki/Logogram)). As for word count, each character counts as one "word", but for keyword analysis, the combination of two or more characters must be seen as one word.

Basically the approach is to take the _word count_ as _character count_, mostly because there are no spaces between characters (in the Chinese case this is what happens).

**Comment by a4jp-com** [comment]
Individual kanji characters sometimes have a meaning but in some situations they are combined with a few other hiragana characters to make different words with different readings.

Kanji only:
下 (shita) down

Kanji with hiragana:
下さい (kudasai) please

Each character in either hiragana, katakana or kanji takes up the space of 2 english characters.

**Comment by terw-dan** [comment]
Thanks both for the explanation. This will come in helpful when we start implementing and testing this.

**Comment by a4jp-com** [comment]
Is there anything I can do to help?

**Comment by IreneStr** [comment]
@a4jp-com Thank you for your eagerness to help! At the moment, our main problem is the absence of spaces in Asian languages. Could you confirm that in Japanese there are no spaces between individual words as well.

**Comment by a4jp-com** [comment]
Sometimes a regular space is used but in other situations a Japanese space is used.

Here is the encoding for the unicode character 'IDEOGRAPHIC SPACE' (U+3000)

**Comment by IreneStr** [comment]
@a4jp-com Thank you for your reply. From googling some Japanese websites, I got the impression that spaces are generally speaking only used between sentences, but not between words. 

In the following sentences, for example, there is a whitespace after 、 and 。.):

> 日本の文化を美術や音楽、演劇、映画からファッションやデザインまで幅広く世界に紹介しています。また、言葉を超えた共感の場をつくり出し、ともに創造する喜びをわかちあって、人と人との交流を深めていきます。

However, these white spaces are part of the 、(U+3001) and 。(U+3002) characters (so the space is not a separate character).

In what situations do people use the regular space or Japanese space?

**Comment by a4jp-com** [comment]
Usually between names
山田 たろ

**Comment by idpokute** [comment]
Most of Japanese don't use space between words. Maybe Yoast can give the option; that turns off some deduction rules for Japanese.

**Comment by a4jp-com** [comment]
I've been living in Japan for 16 years and have worked as a system engineer in 3 Japanese companies. Spaces are used in sites here.

**Comment by IreneStr** [comment]
@idpokute @a4jp-com Thank you both for the information. It'll be very helpful when we want to implement this feature in the future.

**Comment by iamazik** [comment]
Please inform the customer of conversation # 179998 when this conversation has been closed.

**Comment by a4jp-com** [comment]
@IreneStr thank you very much.

**Comment by monbauza** [comment]
Please inform the customer of conversation # 184521 when this conversation has been closed.

**Comment by a4jp-com** [comment]
I have been following the rules set out in the plugin but I have lost 75% of my views on one Japanese site. I've gone from about 200 views a day down to about 50 views a day. 

The only other change I made was changing the site to HTTPS. I thought that was meant to increase the ranking. Any ideas what could be causing the problem?  [https://agreatdream.com/word-lists/](url)

**Comment by ullivr** [comment]
i wrote my blog in chinese, really really really need this function.

**Comment by terw-dan** [comment]
@a4jp-com The wordcount is only shown as an indication. It is not something we (can) save to your post that has influence on your rankings. So it has to be something else that caused a decrease.

**Comment by a4jp-com** [comment]
I was just thinking as the numbers are wrong that when we make pages we might be adding titles that are too long etc

**Comment by saitai0802** [comment]
I write my blog in Chinese and Japanese. There is some example of my post title for you to test.
本願寺，錦市場，八坂神社   <==Chinese only
京阪神八日之旅 <==Chinese only
伏見稻荷大社，けんどん屋，奈良公園 <==Chinese mix Japanese 

Please help to fix it , we do need this amazing function! it keeps telling us our posts are poor makes us sad.
Thank you so so much

**Comment by iamazik** [comment]
Please inform the customer of conversation # 192085 when this conversation has been closed.

**Comment by saitai0802** [comment]
has it been fixed yet? guys

**Comment by a4jp-com** [comment]
I'd love to find out the character count through this plugin.

**Comment by iamazik** [comment]
Please inform the customer of conversation # 198257 when this conversation has been closed.

**Comment by a4jp-com** [comment]
Any ideas on checking the word count?

**Comment by rmarcano** [comment]
Please inform the customer of conversation # 204879 when this conversation has been closed.

**Comment by JanFan** [comment]
I am looking forward to seeing any progress here.

Word Segmentation is a unique problem in languages like Chinese, and I recommend using tools like [Jieba](https://github.com/fxsjy/jieba) for Chinese.

**Comment by a4jp-com** [comment]
Thank you for that info JanFan. Counting characters seems like an easy way to solve the problem.

**Comment by paulovsky** [comment]
I agree that it would be simpler for the development of this feature if it would just count the number of **characters** (instead of words. At least in Chinese language, the common practice is to count characters.

English:  **This is a sentence.** > 4 words
Simplified Chinese: **这是一句话。**> 6 chars (punctuation marks included)

**Comment by a4jp-com** [comment]
Can't the SEO plugin count both words and characters?

**Comment by saitai0802** [comment]
The word count and focus keyword comparison to SEO title is not correct in other language too.

**Comment by paulovsky** [comment]
@saitai0802 Can you please be more specific and give some examples of what you mean?

**Comment by a4jp-com** [comment]
Could someone make a character/word count option in the settings area?

**Comment by 7creo** [comment]
Still no fix for this ... Impressive that the first post about this goes back to 2013. Any chance that this will be implemented? SEO is not exclusive to non-asian languages, right?

**Comment by monbauza** [comment]
Please inform the customer of conversation # 246277 when this conversation has been closed.

**Comment by jacktator** [comment]
Any fix or commits on this issue yet?

**Comment by a4jp-com** [comment]
Any type of fix is greatly appreciated.

**Comment by atimmer** [comment]
@manuelaugustin is going to work on Japanese support of YoastSEO.js. So this will obviously be part of it. He will have updates when he get’s there.

**Comment by a4jp-com** [comment]
Thank you so much. I can help fix bit in Japanese if needed after the 30th.

**Comment by fang-yu** [comment]
Does yoast recognize chinese Focus keyword yet? I have the same problem.

**Comment by a4jp-com** [comment]
Fang-yu, Yoast doesn't count Chinese or Japanese characters properly yet. Hope it will do so soon though.

**Comment by rumejan** [comment]
Please inform the customer of conversation # 281736 when this conversation has been closed.

**Comment by monbauza** [comment]
Please inform the customer of conversation # 282280 when this conversation has been closed.

**Comment by a4jp-com** [comment]
Here is some code to count the Japanese characters: [https://sacha.me/Countable/](url)
[https://github.com/RadLikeWhoa/Countable](url) hope that helps.

**Comment by EricEidel** [comment]
2018 here, still an issue. Really looking forward to have this solved!

**Comment by michaelbriantina** [comment]
Please inform the customer of conversation # 314559 when this conversation has been closed.

**Comment by iamazik** [comment]
Please inform the customer of conversation # 319637 when this conversation has been closed.

**Comment by iamazik** [comment]
Please inform the customer of conversation # 320100 when this conversation has been closed.

**Comment by iamazik** [comment]
Please inform the customer of conversation # 332540 when this conversation has been closed.

**Comment by a4jp-com** [comment]
Any progress on this yet? I just want to check if it is months away or almost fixed if possible.

**Comment by manuelaugustin** [comment]
Thanks for your interest in Japanese langauge support @a4jp-com! We currently have it on our agenda, but it will still take a while before it's going to be implemented.

**Comment by johnsonleee** [comment]
Counting Chinese words is a big problem. First it must have a Chinese words dictionary, then use it to split the text, then count keywords. We should search if there is some Chinese splitting and counting JS library.

**Comment by joe528** [comment]
I don't think you need a dictionary to count Chinese words. You don't need to split words. Each word is already a word. And for keywords, it is just a substring comparison. As someone above already pointed out "At least in the Chinese language, the practice is to count characters."

**Comment by michaelbriantina** [comment]
Please inform the customer of conversation # 372393 when this conversation has been closed.

**Comment by iamazik** [comment]
Please inform the customer of conversation # 374471 when this conversation has been closed.

**Comment by suascat** [comment]
Please inform the customer of conversation # 399452 when this conversation has been closed.

**Comment by a4jp-com** [comment]
A simple character count option sometime would be nice. The code for the word count is there so it shouldn't be too hard to count characters.

**Comment by pyericz** [comment]
Why not try this. [https://github.com/fukuball/jieba-php#jieba-php-english-document](https://github.com/fukuball/jieba-php#jieba-php-english-document)

**Comment by pyericz** [comment]
```Jieba``` is good for Chinese segmentation. It has many languages support. Check out this.
[https://github.com/search?q=jieba](https://github.com/search?q=jieba)

**Comment by Pcosta88** [comment]
Please inform the customer of conversation # 425222 when this conversation has been closed.

**Comment by a4jp-com** [comment]
Can someone make a patch with just a character count?

**Comment by tatitanizaki** [comment]
The character count would help a lot already. 
This issue has been open for more than 2 years. Is this going to be fixed or not?

**Comment by namster2k** [comment]
Microsoft Word counts every Japanese character as a word. This includes hiragana, katakana, and kanji characters. For example, the kanji for "cat" is "猫" and this is counted as one word. The hiragana for "cat" is "ねこ" and this is counted as 2 words. It's not the greatest system, but in my opinion it's better than what Yoast is doing at the moment of counting entire paragraphs as a single word.

**Comment by 7creo** [comment]
Its okay to ignore Chinese and similar languages for years. Yeah, who speaks Chinese anyways… just like 1.6 Billion people...

**Comment by gersangwiki** [comment]
In Korean, we placa a suffix behind a lot of words.  But Yoast is recognizing the words and suffix together as a one keyword.  example 나는 기분이 좋다.  나=I   는=am   기분=feeling  이=is   좋다=good   Let's say my keyword is 기분 (which means feeling) .  But YoastSEO thinks the keyword is '기분이' because they are together.  In English, this means that Yoast is recognizing 'feeling is' as a keyword instead of 'feeling' 
Recognizing only 6 suffixes separately with the word will improve this a lot.
Most used Korean suffixes : 을or를, 이or가, 은or는.
If it's possible, try to make it like this > When one of 을,를,이,가,은,는 is placed at the end, it's used as a suffix, so make yoast recognize the keyword except the last character.  However, when it's place at the front or in the middle, it's part of the keyword.
Fixing this will be crucial for Korean users.

**Comment by gersangwiki** [comment]
In Korean, we place a suffix behind a lot of words. But Yoast is recognizing the words and suffix together as a one keyword. example 나는 기분이 좋다. 나=I 는=am 기분=feeling 이=is 좋다=good Let's say my keyword is 기분 (which means feeling) . But YoastSEO thinks the keyword is '기분이' because they are together. In English, this means that Yoast is recognizing 'feeling is' as a keyword instead of 'feeling'
Recognizing only 6 suffixes separately with the word will improve this a lot.
Most used Korean suffixes : 을or를, 이or가, 은or는.
If it's possible, try to make it like this > When one of 을,를,이,가,은,는 is placed at the end, it's used as a suffix, so make yoast recognize the keyword except the last character. However, when it's place at the front or in the middle, it's part of the keyword.
Fixing this will be crucial for Korean users.

**Comment by a4jp-com** [comment]
Is just having a character count good for the Korean language?

**Comment by topout** [comment]
@moorscode said 

> Need-spec needed from a linguisticsteam perspective

There're many Asian languages. While they are similar, there are some differences so I suggest let's start with Chinese because it is the most spoken language (and I use it :D ) and seems easier than some of the other languages. 

The below suggested specs can be applied to both Simplified Chinese (learned by people living in Mainland China) and Traditional Chinese (learned by people who learned Chinese outside Mainland China). 

My suggested Spec for Chinese (Simplified and Traditional Chinese):

- We need to count each character, whether there's a space or not, as an individual word.
- If there are special characters, such as periods, parenthesis, etc., treat it like you would if those were English words. Chinese uses parenthesis, commas, etc. like English grammar and it is similar for most cases.
- In case Yoast handles periods to end a sentence differently, Chinese uses "。" in place of a ". " to end a sentence. 

If we can do that, at least targeting keywords, keyphrase, synonyms, will work which will make Yoast much more useful as right now where it just does not work at all. Things like Flesch Reading, Passive voice, Transition words won't work but I think many people, including myself, will still be grateful at least we can focus on keywords.

**Comment by 7creo** [comment]
Yes, you are absolutely correct.
A word count would already be a big achievement. But for 2 years since I first read about this issue, nothing has happened. And I don’t think it is that hard to implement.

**Comment by laurasacco** [comment]
Please inform the customer of conversation # 453293 when this conversation has been closed.

**Comment by iamazik** [comment]
Please inform the customer of conversation # 453634 when this conversation has been closed.

**Comment by iamazik** [comment]
Please inform the customer of conversation # 457569 when this conversation has been closed.

**Comment by jerclarke** [comment]
This sad thread has nothing to do with me, but I stumbled onto it while trying to figure out how to re-implement the core word-counting JS from WP and I found a solution:

https://wordpress.stackexchange.com/questions/279152/can-i-make-use-of-word-count-js-in-my-own-code

The neat thing is that core WP ships different versions of `word-count.js` depending on your locale, so e.g. if you use `window.wp.utils.WordCounter()` with Chinese localization, it counts characters instead of words. 

Seems like implementing it this way is the obvious choice for just about any plugin, since you can offload all the research and dealing with users to core WP's choices (which are probably pretty good!)

Good luck!

**Comment by a4jp-com** [comment]
Fantastic info @jerclarke 
I hope this type of code can be added as soon as possible. 
This conversation has been running over 2 years and there are now even samples of code that can fix the problem. I just hope someone in the company puts a few people on this finally. Why doesn't someone just add a few if statements to the code? The calculation code is already in the plugin.

**Comment by laurasacco** [comment]
Please inform the customer of conversation # 489162 when this conversation has been closed.

**Comment by Pcosta88** [comment]
Please inform the customer of conversation # 411495 when this conversation has been closed.

**Comment by iamazik** [comment]
Please inform the customer of conversation # 502775 when this conversation has been closed.

**Comment by suascat** [comment]
Please inform the customer of conversation # 511434 when this conversation has been closed.

**Comment by maybellyne** [comment]
Please inform the customer of conversation # 515980 when this conversation has been closed

**Comment by a4jp-com** [comment]
Is anyone still working on this?

**Comment by amboutwe** [comment]
Please inform the customer of conversation # 516867 when this conversation has been closed.

**Comment by amboutwe** [comment]
Please inform the customer of conversation # 538393 when this conversation has been closed.

**Comment by hawm** [comment]
Still a problem at Sep 2019.

**Comment by mayada-ibrahim** [comment]
Please inform the customer of conversation # 565766 when this conversation has been closed.

**Comment by iamazik** [comment]
Please inform the customer of conversation # 585879 when this conversation has been closed.

**Comment by a4jp-com** [comment]
Is anyone working on this anymore?

**Comment by Djennez** [comment]
This is currently not being worked on, it is a feature request that might get implemented in the future.

**Comment by a4jp-com** [comment]
Okay. Thanks for the honest reply @Djennez. I'll post this update in the plugin download area of WordPress. No one has said this up till now which isn't very good. This should have been made clear years ago. 

Is it okay to fork the current plugin and make a Japanese version. I'll just edit the character count code. What is the licence of the current free version? Is it a General Public License?

**Comment by Pcosta88** [comment]
Please inform the customer of conversation # 598441 when this conversation has been closed.

**Comment by 7creo** [comment]
> Okay. Thanks for the honest reply @Djennez. I'll post this update in the plugin download area of WordPress. No one has said this up till now which isn't very good. This should have been made clear years ago.
> 
> Is it okay to fork the current plugin and make a Japanese version. I'll just edit the character count code. What is the licence of the current free version? Is it a General Public License?

Did you success on this yet? I still don't understand how this situation can be ignored by the developers for such a long time. I would never purchase a paid version of that with this issue, as it makes the plugin 50% useless.

**Comment by Pcosta88** [comment]
Please inform the customer of conversation # 607816 when this conversation has been closed.

**Comment by michaelbriantina** [comment]
https://wordpress.org/support/topic/cjk-word-count/

**Comment by a4jp-com** [comment]
Can you include a conditional statement based on the page language that either counts words or characters based on the language set?  

Example 1:

`var lang = document.documentElement.lang
if (lang === ‘ja’) {
...do something - Japanese character count code here...
}`

Example 2:

`<?php if(ICL_LANGUAGE_CODE=='zh' || ICL_LANGUAGE_CODE=='ja' || ICL_LANGUAGE_CODE=='ko'){ ?>
...Chinese/Japanese/Korean character count code...
<?php } ?>`

I'm not sure how you have separated other languages but I'm sure you already have code like this in your plugin for other languages.

**Comment by crpeng** [comment]
I also need this improvement for Chinese language.

**Comment by a4jp-com** [comment]
T-T. Whatever happened to the programmer that was assigned to fixing this problem?

**Comment by suascat** [comment]
Please inform the customer of conversation # 649514 when this conversation has been closed.

**Comment by priscillamc** [comment]
https://wordpress.org/support/topic/seo-in-japanese/

**Comment by michaelbriantina** [comment]
Please inform the customer of conversation # 729318 when this conversation has been closed.

**Comment by a4jp-com** [comment]
Is there a way to just get a character count instead of word count? You already have the code in the plugin. You just need a function for the option.

**Comment by ogodoabiola** [comment]
Please inform the customer of conversation #744721 when this conversation has been closed.

**Comment by Isildur00** [comment]
Please inform the customer of conversation #760565 when this conversation has been closed.

**Comment by a4jp-com** [comment]
> WordPress has build code to be able to count words in Asian languages. We should take this as inspiration. WordPress actually ships the word counting code only in the Chinese language pack, the third link shows where to find it.
> 
> Patch for word-count.js
> https://core.trac.wordpress.org/ticket/20738
> https://core.trac.wordpress.org/ticket/30966
> 
> Where to find the language pack:
> https://core.trac.wordpress.org/ticket/33454

Can we find out what is happening if you already have the code please?
